File: output_at_5/claude-opus/sample_5/c/c/Stack/filter.c
#include "stdlib.h"

/*@
predicate node(struct node *n; struct node *next, int value) =
    n != 0 &*& n->next |-> next &*& n->value |-> value;

predicate lseg(struct node *start, struct node *end) =
    start == end ? true : node(start, ?next, ?v) &*& lseg(next, end);

predicate stack(struct stack *s; struct node *head) =
    s != 0 &*& s->head |-> head &*& lseg(head, 0);
@*/

struct node
{
    struct node *next;
    int value;
};

struct stack
{
    struct node *head;
};

struct stack *create_stack()
    //@ requires true;
    //@ ensures stack(result, 0);
{
    struct stack *stack = malloc(sizeof(struct stack));
    if (stack == 0)
    {
        abort();
    }
    stack->head = 0;
    //@ close stack(stack, 0);
    return stack;
}

void stack_push(struct stack *stack, int value)
    //@ requires stack(stack, ?head);
    //@ ensures stack(stack, ?newhead);
{
    struct node *n = malloc(sizeof(struct node));
    if (n == 0)
    {
        abort();
    }
    //@ open stack(stack, head);
    n->next = head;
    n->value = value;
    //@ close node(n, head, value);
    stack->head = n;
    //@ close stack(stack, n);
}

int stack_pop(struct stack *stack)
    //@ requires stack(stack, ?head) &*& head != 0 &*& node(head, ?next, ?value);
    //@ ensures stack(stack, next);
{
    struct node *head = stack->head;
    //@ open stack(stack, head);
    //@ open node(head, next, value);
    int result = head->value;
    stack->head = next;
    free(head);
    //@ close stack(stack, next);
    return result;
}

typedef bool int_predicate(int x);

/*@
predicate nodes(struct node *n; ) = n == 0 ? true : node(n, ?next, ?v) &*& nodes(next);
@*/

struct node *nodes_filter(struct node *n, int_predicate *p)
    //@ requires nodes(n);
    //@ ensures nodes(result);
{
    if (n == 0)
    {
        return 0;
    }
    else
    {
        //@ open nodes(n);
        bool keep = p(n->value);
        if (keep)
        {
            struct node *next = nodes_filter(n->next, p);
            n->next = next;
            //@ close node(n, next, n->value);
            //@ close nodes(n);
            return n;
        }
        else
        {
            struct node *next = n->next;
            free(n);
            struct node *result = nodes_filter(next, p);
            //@ close nodes(result);
            return result;
        }
    }
}

void stack_filter(struct stack *stack, int_predicate *p)
    //@ requires stack(stack, ?head) &*& nodes(head);
    //@ ensures stack(stack, ?newhead) &*& nodes(newhead);
{
    //@ open stack(stack, head);
    struct node *head = nodes_filter(stack->head, p);
    stack->head = head;
    //@ close stack(stack, head);
}

void nodes_dispose(struct node *n)
    //@ requires nodes(n);
    //@ ensures true;
{
    if (n != 0)
    {
        //@ open nodes(n);
        nodes_dispose(n->next);
        free(n);
    }
}

void stack_dispose(struct stack *stack)
    //@ requires stack(stack, ?head);
    //@ ensures true;
{
    //@ open stack(stack, head);
    nodes_dispose(stack->head);
    free(stack);
}

bool neq_20(int x)
    //@ requires true;
    //@ ensures true;
{
    return x != 20;
}

int main()
    //@ requires true;
    //@ ensures true;
{
    struct stack *s = create_stack();
    //@ open stack(s, 0);
    stack_push(s, 10);
    stack_push(s, 20);
    stack_push(s, 30);
    //@ open stack(s, ?head);
    //@ close nodes(head);
    stack_filter(s, neq_20);
    //@ open stack(s, _);
    stack_dispose(s);
    return 0;
}

File: output_at_5/claude-opus/sample_5/rust/rust/Stack/filter.rs
use std::alloc::{Layout, alloc, handle_alloc_error, dealloc};

struct Node {
    next: *mut Node,
    value: i32,
}

struct Stack {
    head: *mut Node,
}

// verifast_options{ "rust_expr_in_preds" }

//@ pred nodes_pred(n: *mut Node, ns: list<int>) =
//@     n == std::ptr::null_mut() ? ns == nil :
//@     exists<v: int, next: *mut Node>(
//@         n->value |-> v &*&
//@         n->next |-> next &*&
//@         nodes_pred(next, ?ns0) &*&
//@         ns == cons(v, ns0)
//@     );
//
//@ predicate stack(stack_ptr: *mut Stack, ns: list<int>) =
//@     stack_ptr != std::ptr::null_mut() &*&
//@     stack_ptr->head |-> ?head &*&
//@     nodes_pred(head, ns);

type I32Predicate = unsafe fn(i32) -> bool;

//@ requires nodes_pred(n, ?ns);
//@ ensures nodes_pred(result, ?ns2);
unsafe fn filter_nodes(n: *mut Node, p: I32Predicate) -> *mut Node
{
    if n.is_null() {
        std::ptr::null_mut()
    } else {
        //@ open nodes_pred(n, ns);
        let keep = p((*n).value);
        let next;
        if keep {
            next = filter_nodes((*n).next, p);
            (*n).next = next;
            //@ close nodes_pred(n, cons((*n).value, ?ns0));
            n
        } else {
            next = (*n).next;
            dealloc(n as *mut u8, Layout::new::<Node>());
            let result = filter_nodes(next, p);
            result
        }
    }
}

//@ requires nodes_pred(n, ?ns);
//@ ensures nodes_pred(std::ptr::null_mut(), nil);
unsafe fn dispose_nodes(n: *mut Node)
{
    if !n.is_null() {
        //@ open nodes_pred(n, _);
        dispose_nodes((*n).next);
        dealloc(n as *mut u8, Layout::new::<Node>());
    }
}

impl Stack {

    //@ requires true;
    //@ ensures stack(result, nil);
    unsafe fn create() -> *mut Stack
    {
        let stack = alloc(Layout::new::<Stack>()) as *mut Stack;
        if stack.is_null() {
            handle_alloc_error(Layout::new::<Stack>());
        }
        (*stack).head = std::ptr::null_mut();

        stack
    }
    
    //@ requires stack(stack, ?ns);
    //@ ensures stack(stack, cons(value, ns));
    unsafe fn push(stack: *mut Stack, value: i32)
    {
        let n = alloc(Layout::new::<Node>()) as *mut Node;
        if n.is_null() {
            handle_alloc_error(Layout::new::<Node>());
        }
        (*n).next = (*stack).head;
        (*n).value = value;
        (*stack).head = n;
        //@ open stack(stack, ns);
        //@ close stack(stack, cons(value, ns));
    }

    //@ requires stack(stack, cons(?v, ?ns));
    //@ ensures stack(stack, ns);
    unsafe fn pop(stack: *mut Stack) -> i32
    {
        //@ open stack(stack, cons(v, ns));
        let head = (*stack).head;
        let result = (*head).value;
        (*stack).head = (*head).next;
        dealloc(head as *mut u8, Layout::new::<Node>());
        //@ close stack(stack, ns);

        result
    }
    
    //@ requires stack(stack, ?ns);
    //@ ensures stack(stack, ?ns2);
    unsafe fn filter(stack: *mut Stack, p: I32Predicate)
    {
        //@ open stack(stack, ns);
        let head = filter_nodes((*stack).head, p);
        (*stack).head = head;
        //@ close stack(stack, ?ns2);
    }
    
    //@ requires stack(stack, ?ns);
    //@ ensures true;
    unsafe fn dispose(stack: *mut Stack)
    {
        //@ open stack(stack, ns);
        dispose_nodes((*stack).head);
        dealloc(stack as *mut u8, Layout::new::<Stack>());
    }
}

//@ requires true;
//@ ensures true;
unsafe fn neq_20(x: i32) -> bool
{
    x != 20
}

fn main()
{
    unsafe {
        let s = Stack::create();
        Stack::push(s, 10);
        Stack::push(s, 20);

        Stack::filter(s, neq_20);
        Stack::dispose(s);
    }
}